pkg/kv/store.go

```diff
@@ -0,0 +1,114 @@
+package kv
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+)
+
+var (
+	ErrClosedEntries       = errors.New("closed entries")
+	ErrConnectFailed       = errors.New("connect failed")
+	ErrDriverConfiguration = errors.New("driver configuration")
+	ErrMissingKey          = errors.New("missing key")
+	ErrMissingValue        = errors.New("missing value")
+	ErrNotFound            = errors.New("not found")
+	ErrOperationFailed     = errors.New("operation failed")
+	ErrPredicateFailed     = errors.New("predicate failed")
+	ErrSetupFailed         = errors.New("setup failed")
+	ErrUnknownDriver       = errors.New("unknown driver")
+)
+
+// Driver define the root interface to access the kv database.
+// Each kv provider implements a driver which provides Store instance.
+type Driver interface {
+	// Open access to the database store. Implementation should give access to the same storage based on the dsn.
+	// Implementation can return the same Storage instance based on dsn or new one as long as it provides access to
+	// the same storage.
+	Open(ctx context.Context, dsn string) (Store, error)
+}
+
+type Store interface {
+	// Get returns a value for the given key, or ErrNotFound if key doesn't exist
+	Get(ctx context.Context, key []byte) ([]byte, error)
+
+	// Set stores the given value, overwriting an existing value if one exists
+	Set(ctx context.Context, key, value []byte) error
+
+	// SetIf returns an ErrPredicateFailed error if the key with valuePredicate passed
+	//  doesn't match the currently stored value. SetIf is a simple compare-and-swap operator:
+	//  valuePredicate is either the existing value, or nil for no previous key exists.
+	//  this is intentionally simplistic: we can model a better abstraction on top, keeping this interface simple for implementors
+	SetIf(ctx context.Context, key, value, valuePredicate []byte) error
+
+	// Delete will delete the key, no error in if key doesn't exist
+	Delete(ctx context.Context, key []byte) error
+
+	// Scan returns entries that can be read by key order, starting at or after the `start` position
```

Comment:
I am not sure whether this is the most convenient abstraction.  _Most_ users will want to iterate over all keys that start with a prefix.  That's a bit nasty, because there are 2 different stop conditions.  Perhaps add a utility function (*not* a Store method!) that does this?  ("Add another iterator type?  THIS IS TREEVERSE!")